### PR TITLE
JSON output for cortex get

### LINKED ISSUE #1161: Support json output for CLI commands
#### Description

Add a flag (e.g. `--json`) to certain CLI commands to output the response in structured json format.

#### Questions

* Which commands should support this? Definitely `get`; any others?

_suggested by https://github.com/cortexlabs/cortex/issues/1160_

### PR DESCRIPTION
**Pull request by vishalbollu**
closes #1161 

---

checklist:

- [ ] run `make test` and `make lint`
- [ ] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)

### CODE REVIEW

File: cli/types/flags/output_type.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+func (t *OutputType) UnmarshalBinary(data []byte) error {
+	return t.UnmarshalText(data)
+}
+
+// MarshalBinary satisfies BinaryMarshaler
+func (t OutputType) MarshalBinary() ([]byte, error) {
+	return []byte(t.String()), nil
+}
+
+func (t *OutputType) Set(value string) error {
```
Comment (deliahu) [inline cli/types/flags/output_type.go:80]: I don't remember seeing `Set()` and `Type()` in the other enums that we have, what is it used for and is it worth adding it to the rest? (could be a separate PR if yes)

File: cli/types/flags/errors.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"fmt"
+
+	"github.com/cortexlabs/cortex/pkg/lib/errors"
+	s "github.com/cortexlabs/cortex/pkg/lib/strings"
+)
+
+const (
+	ErrInvalidOutputType = "cli.invalid_output_type"
```
Comment (deliahu) [inline cli/types/flags/errors.go:27]: Should be "flags.invalid_output_type"

File: cli/types/flags/errors.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+func ErrorInvalidOutputType(invalidOutputType string, validOutputTypes []string) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrInvalidOutputType,
+		Message: fmt.Sprintf("invalid value specified for flag --output %s, valid output values are %s", invalidOutputType, s.StrsAnd(validOutputTypes)),
```
Comment (deliahu) [inline cli/types/flags/errors.go:33]: (nit) how does this sound to you?

```go
fmt.Sprintf("invalid value \"%s\" specified for -o/--output; valid values are %s", invalidOutputType, s.UserStrsAnd(validOutputTypes)),
```

File: cli/cmd/errors.go [hunk 4]
```diff
@@ -79,6 +82,20 @@ func ErrorInvalidProvider(providerStr string) error {
 	})
 }
 
+func ErrorInvalidOutputType(invalidOutputType string, validOutputTypes []string) error {
```
Comment (deliahu) [inline cli/cmd/errors.go:85]: This is not used

File: cli/types/flags/errors.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+	"github.com/cortexlabs/cortex/pkg/lib/errors"
+	s "github.com/cortexlabs/cortex/pkg/lib/strings"
+)
+
+const (
+	ErrInvalidOutputType = "cli.invalid_output_type"
+)
+
+func ErrorInvalidOutputType(invalidOutputType string, validOutputTypes []string) error {
```
Comment (deliahu) [inline cli/types/flags/errors.go:30]: nit: `validOutputTypes` does not need to be passed in

File: cli/cmd/lib_batch_apis.go [hunk 6]
```diff
@@ -143,6 +145,14 @@ func getJob(env cliconfig.Environment, apiName string, jobID string) (string, er
 		return "", err
 	}
 
+	if _flagGetOutput == flags.JSONOutputType {
+		bytes, err := json.Marshal(resp)
+		if err != nil {
+			return "", err
```
Comment (deliahu) [inline cli/cmd/lib_batch_apis.go:151]: the error should be wrapped, or we can use libjson

File: cli/cmd/root.go [hunk 7]
```diff
@@ -176,6 +177,19 @@ func updateRootUsage() {
 	})
 }
 
+func shouldPrintHelper(cmd *cobra.Command) bool {
```
Comment (deliahu) [inline cli/cmd/root.go:180]: can be remove by use of shared flag

File: cli/cmd/get.go [hunk 8]
```diff
@@ -157,6 +174,13 @@ func getAPIsInAllEnvironments() (string, error) {
 	var allTrafficSplitters []schema.TrafficSplitter
 	var allTrafficSplitterEnvs []string
 
+	type getAPIsOutput struct {
+		APIs  schema.GetAPIsResponse
+		Error string
+	}
+
+	allAPIsOutput := map[string]getAPIsOutput{}
```
Comment (deliahu) [inline cli/cmd/get.go:182]: nit: add an inline comment, e.g.

```go
allAPIsOutput := map[string]getAPIsOutput{} // env name -> api output
```

File: cli/cmd/get.go [hunk 9]
```diff
@@ -182,8 +210,22 @@ func getAPIsInAllEnvironments() (string, error) {
...
 			errorsMap[env.Name] = err
 		}
+
+		allAPIsOutput[env.Name] = apisOutput
+	}
+
+	if _flagOutput == flags.JSONOutputType {
+		bytes, err := libjson.Marshal(allAPIsOutput)
+		if err != nil {
+			if err != nil {
```
Comment (deliahu) [inline cli/cmd/get.go:223]: this line is unnecessary

File: cli/cmd/get.go (hunk 8 above)
Comment (deliahu) [reply cli/cmd/get.go:182]: edit: I think it might be better to instead add `Env string` to the `getAPIsOutput` struct, and to serialize `[]getAPIsOutput` instead of the map. lmk if you want to discuss

File: cli/cmd/errors.go [hunk 10]
```diff
@@ -79,6 +81,13 @@ func ErrorInvalidProvider(providerStr string) error {
 	})
 }
 
+func ErrorEnvironmentFlagRequired(outputType flags.OutputType) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrEnvironmentFlagRequired,
+		Message: fmt.Sprintf("'--output %s' is only supported when the environment flag is provided (e.g. cortex get --env <env_name>)", outputType.String()),
```
[the remark this answers is not in the document]
Comment (vishalbollu) [reply cli/cmd/errors.go:87]: This error is not relevant anymore and will be deleted.

File: cli/cmd/get.go (hunk 8 above)
Comment (vishalbollu) [reply cli/cmd/get.go:182]: Great Idea

File: cli/cmd/root.go [hunk 11]
```diff
@@ -193,7 +195,9 @@ func printEnvIfNotSpecified(envName string, cmd *cobra.Command) error {
 		return err
 	}
 
-	fmt.Print(out)
+	if _flagOutput == flags.PrettyOutputType {
```
[the remark this answers is not in the document]
Comment (vishalbollu) [reply cli/cmd/root.go:198]: Will leave it for now based on our discussions.

File: cli/cmd/root.go [hunk 12]
```diff
@@ -203,7 +207,7 @@ func envStringIfNotSpecified(envName string, cmd *cobra.Command) (string, error)
 		return "", err
 	}
 
-	if !wasEnvFlagProvided(cmd) && len(envNames) > 1 {
+	if _flagOutput == flags.PrettyOutputType && !wasEnvFlagProvided(cmd) && len(envNames) > 1 {
```
[the remark this answers is not in the document]
Comment (vishalbollu) [reply cli/cmd/root.go:210]: Will leave it for now based on our discussions.

File: pkg/lib/json/json.go [hunk 13]
```diff
@@ -25,14 +25,22 @@ import (
...
-func Marshal(obj interface{}) ([]byte, error) {
+func MarshalIndent(obj interface{}) ([]byte, error) {
 	jsonBytes, err := json.MarshalIndent(obj, "", "  ")
 	if err != nil {
 		return nil, errors.Wrap(err, errStrMarshalJSON)
 	}
 	return jsonBytes, nil
 }
 
+func Marshal(obj interface{}) ([]byte, error) {
```
Comment (deliahu) [inline pkg/lib/json/json.go:36]: Because of this change, it will be necessary to search for all occurrences of `json.Marshal` (where `json` is our json package) and replace them with `MarshalIndent()`

Comment (vishalbollu) [reply pkg/lib/json/json.go:36]: I did a pass where I was going to replace the previous occurrences of `json.Marshal` with `json.MarshalIndent` but I couldn't find any instances where they were user facing so I didn't update them.

**Review by deliahu — APPROVED** [review]
